app/workers/tasks.py

```diff
@@ -38,66 +38,77 @@ async def test_task(message: str) -> None:
-            fast_to_checksum_address(address), chain_id
-        )
-        result = await contract_metadata_service.process_contract_metadata(
-            session, contract_metadata
-        )
-        if result:
-            logger.info(
-                "Success download contract metadata for contract=%s and chain=%s",
-                address,
-                chain_id,
+    address: str, chain_id: int, skip_attemp_download: bool = False
+):
+    async with AsyncSession(get_engine(), expire_on_commit=False) as session:
+        contract_metadata_service = get_contract_metadata_service()
+        # Just try the first time, following retries should be scheduled
+        if (
+            skip_attemp_download
+            or await contract_metadata_service.should_attempt_download(
+                session, address, chain_id, 0
             )
-        else:
+        ):
             logger.info(
-                "Failed to download contract metadata for contract=%s and chain=%s",
+                "Downloading contract metadata for contract=%s and chain=%s",
                 address,
                 chain_id,
             )
+            contract_metadata = await contract_metadata_service.get_contract_metadata(
+                fast_to_checksum_address(address), chain_id
+            )
+            result = await contract_metadata_service.process_contract_metadata(
+                session, contract_metadata
+            )
+            if result:
+                logger.info(
+                    "Success download contract metadata for contract=%s and chain=%s",
+                    address,
+                    chain_id,
+                )
+            else:
+                logger.info(
+                    "Failed to download contract metadata for contract=%s and chain=%s",
+                    address,
+                    chain_id,
+                )
+
+            if proxy_implementation_address := contract_metadata_service.get_proxy_implementation_address(
+                contract_metadata
+            ):
+                logger.info(
+                    "Adding task to download proxy implementation metadata from address=%s for contract=%s and chain=%s",
+                    proxy_implementation_address,
+                    address,
+                    chain_id,
+                )
+                get_contract_metadata_task.send(
+                    address=proxy_implementation_address, chain_id=chain_id
+                )
+        else:
+            logger.debug("Skipping contract=%s and chain=%s", address, chain_id)
 
-        if proxy_implementation_address := contract_metadata_service.get_proxy_implementation_address(
-            contract_metadata
+
+@dramatiq.actor(periodic=cron("0 0 * * *"))  # Every midnight
+async def get_missing_contract_metadata_task():
+    async with AsyncSession(get_engine(), expire_on_commit=False) as session:
+        async for contract in Contract.get_contracts_without_abi(
```

Comment:
@falvaradorodriguez @Uxio0 I'm removing database session decorator here. 
I'm thinking that maybe would be better to move all the Sessions to the model file, with the current approach if a session have any issue in the middle the task will fail. Maybe we can open a session by query instead by task. 
What do you think? We could handle it in a different PR.

Reply:
Not sure about this. I feel in this case is alright, but we should address the rest in a different PR